From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: PureGero <puregero@gmail.com>
Date: Wed, 5 Jan 2022 19:49:32 +1000
Subject: [PATCH] Allow for only ip forwarding with no uuid

Allow proxies to forward only the IP without a UUID by accepting a two
part split in the login hostname, and only reading the spoofed UUID when
it is present.

diff --git a/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java b/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
index 649f964..991f24a 100644
--- a/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
+++ b/net/minecraft/server/network/ServerHandshakePacketListenerImpl.java
@@ -153,12 +153,13 @@ public class ServerHandshakePacketListenerImpl implements ServerHandshakePacketL
             String[] split = packet.hostName().split("\00");
             if (!handledByEvent && proxyLogicEnabled) { // Paper
                 // if (org.spigotmc.SpigotConfig.bungee) { // Paper - comment out, we check above!
-                if ((split.length == 3 || split.length == 4) && (ServerHandshakePacketListenerImpl.BYPASS_HOSTCHECK || ServerHandshakePacketListenerImpl.HOST_PATTERN.matcher(split[1]).matches())) { // Paper - Add bypass host check
+                if ((split.length == 2 || split.length == 3 || split.length == 4) && (ServerHandshakePacketListenerImpl.BYPASS_HOSTCHECK || ServerHandshakePacketListenerImpl.HOST_PATTERN.matcher(split[1]).matches())) { // Paper - Add bypass host check // MultiPaper - Allow for only ip-forwarding
                     // Paper start - Unix domain socket support
                     java.net.SocketAddress socketAddress = this.connection.getRemoteAddress();
                     this.connection.hostname = split[0];
                     this.connection.address = new java.net.InetSocketAddress(split[1], socketAddress instanceof java.net.InetSocketAddress ? ((java.net.InetSocketAddress) socketAddress).getPort() : 0);
                     // Paper end - Unix domain socket support
+                    if (split.length > 2) // MultiPaper - Allow for only ip-forwarding
                     this.connection.spoofedUUID = com.mojang.util.UndashedUuid.fromStringLenient(split[2]);
                 } else {
                     Component message = Component.literal("If you wish to use IP forwarding, please enable it in your BungeeCord config as well!");
@@ -169,7 +170,7 @@ public class ServerHandshakePacketListenerImpl implements ServerHandshakePacketL
                 if (split.length == 4) {
                     this.connection.spoofedProfile = ServerHandshakePacketListenerImpl.gson.fromJson(split[3], com.mojang.authlib.properties.Property[].class);
                 }
-            } else if ((split.length == 3 || split.length == 4) && (ServerHandshakePacketListenerImpl.HOST_PATTERN.matcher(split[1]).matches())) {
+            } else if ((split.length == 2 || split.length == 3 || split.length == 4) && (ServerHandshakePacketListenerImpl.HOST_PATTERN.matcher(split[1]).matches())) { // MultiPaper - Allow for only ip-forwarding
                 Component message = Component.literal("Unknown data in login hostname, did you forget to enable BungeeCord in spigot.yml?");
                 this.connection.send(new ClientboundLoginDisconnectPacket(message));
                 this.connection.disconnect(message);
